File: frontend/src/types/trace.ts
export type HopType =
  | 'source' | 'dns' | 'kube_proxy' | 'load_balance'
  | 'cni_same_node' | 'cni_cross_node'
  | 'ingress_controller' | 'netpol_check' | 'destination';

export type HopStatus = 'theoretical' | 'allowed' | 'blocked' | 'unknown';

export interface TraceHop {
  step:      number;
  hop_type:  HopType;
  title:     string;
  subtitle:  string;
  detail:    string;
  technical: string;
  node_id:   string | null;
  status:    HopStatus;
}

export interface BackendCandidate {
  pod_id:    string;
  pod_name:  string;
  pod_ip:    string | null;
  node_name: string | null;
  same_node: boolean;
  ready:     boolean;
}

export interface TraceResult {
  from_pod_id:      string;
  to_service_id:    string;
  hops:             TraceHop[];
  candidates:       BackendCandidate[];
  cni_type:         string | null;
  kube_proxy_mode:  string | null;
  dns_fqdn:         string | null;
  cross_node:       boolean;
  has_netpol:       boolean;
  error:            string | null;
}

export interface TraceRequest {
  fromPodId:    string;
  toServiceId:  string;
  fromLabel:    string;
  toLabel:      string;
}

File: backend/app/models/trace.py
from __future__ import annotations
from enum import Enum
from typing import Literal
from pydantic import BaseModel, Field


class HopType(str, Enum):
    SOURCE             = "source"
    DNS                = "dns"
    KUBE_PROXY         = "kube_proxy"
    LOAD_BALANCE       = "load_balance"
    CNI_SAME_NODE      = "cni_same_node"
    CNI_CROSS_NODE     = "cni_cross_node"
    INGRESS_CONTROLLER = "ingress_controller"
    NETPOL_CHECK       = "netpol_check"
    DESTINATION        = "destination"


class HopStatus(str, Enum):
    THEORETICAL = "theoretical"
    ALLOWED     = "allowed"
    BLOCKED     = "blocked"
    UNKNOWN     = "unknown"


class TraceHop(BaseModel):
    step       : int
    hop_type   : HopType
    title      : str
    subtitle   : str
    detail     : str
    technical  : str
    node_id    : str | None = None
    status     : HopStatus = HopStatus.THEORETICAL


class BackendCandidate(BaseModel):
    pod_id    : str
    pod_name  : str
    pod_ip    : str | None
    node_name : str | None
    same_node : bool
    ready     : bool


class TraceResult(BaseModel):
    from_pod_id  : str
    to_service_id: str
    hops         : list[TraceHop]
    candidates   : list[BackendCandidate]
    cni_type     : str | None
    kube_proxy_mode : str | None
    dns_fqdn     : str | None
    cross_node   : bool
    has_netpol   : bool
    error        : str | None = None
